usb: netusb: Rework netusb/ecm

loicpoulain opened the pull request:
- Remove eth_emu
- Add netusb_recv method
- Remove multi-function array
- Abstract netusb function specific methods
- Move endpoint specific config to function interface
- Remove ecm_register_function
- remove ecm struct 

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

finikorg commented on subsys/usb/class/netusb/netusb.c:
```diff
@@ -21,88 +21,69 @@
 	case 1:
-		netusb.conf_index = 0;
-		eth_emu_up();
-		netusb_connect_media();
```
connect_media() callback is needed for RNDIS and probably other protocol, is it needed for EEM? Are you developing EEM at the moment?

loicpoulain replied:
yes I do, but wanted to avoid introducing unused cb, could come with RNDIS or EEM. Since this cb was already present I can probably keep it.

finikorg commented on subsys/usb/class/netusb/netusb.c:
```diff
@@ -232,67 +203,82 @@ int try_write(u8_t ep, u8_t *data, u16_t len)
 
-	return ret;
+static struct netusb_dev netusb0 = {
+	/* Use RFC7042 Documentation for MAC address */
```
why you name it netusb0? are you going to somehow use netusb1, etc?

loicpoulain replied:
who knows ? ;) this is to highlight that this struct is linked to the instance of the device we create.

finikorg replied:
IMO even with multiple configurations we always have only one instance of this structure, so just name it netusb to make code consistent.

finikorg commented on subsys/usb/class/netusb/function_ecm.c:
```diff
@@ -271,27 +267,42 @@ int ecm_send(struct net_pkt *pkt)
 };
 
-struct netusb_function *ecm_register_function(struct net_if *iface, u8_t in)
+struct netusb_function *ecm_register_function(void)
```
I was thinking to make it more general, like register_function(type), etc.

finikorg reviewed (changes requested):
Please send a patch fixing memory leak as a separate patch.

loicpoulain commented:
@finikorg, my bad, memory leak was introduced when I removed eth_emu.

loicpoulain commented:
Comments addressed.

jukkar commented:
Could you split the PR into more logical pieces? I know that it can be hard sometimes but you have now lot of changes described in PR that do not seem to be that much related to each other.

Commit:
usb: netusb: Remove eth_emu

Move network device creation/mgmt to netusb.
Add iface reference to netusb instance.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
usb: netusb: Add netusb_recv method

This method can be use by netusb functions on pkt reception.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
usb: netusb: Remove multi-function array

Today, we support/use only one USB configuration descriptor.
Moreover I assume multi-config support should be managed at
usb core level and not by each class driver.
Let's track one netusb function per netusb instance for now.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
usb: netusb: Abstract netusb function specific methods

Move class_handler and send_pkt to netusb function interface.
This makes netusb 'function' agnostic.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
usb: netusb: Move endpoint specific config to function interface

netusb endpoint config is specific to the function (ECM, RNDIS, EEM..).
Move this config to the function interface.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
usb: netusb: Remove ecm_register_function

The only access to the function is now the function interface.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
usb: netusb: remove ecm struct

Remove unused iface member and move out skip variable.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

loicpoulain commented:
Addressed @jukkar comment, patch split.

jfischer-no reviewed (approved):
LGTM

jukkar commented on subsys/usb/class/netusb/netusb.c:
```diff
@@ -21,27 +21,27 @@
 
 static struct __netusb {
+	struct net_if *iface;
 	u8_t conf;
```
I know the conf variable is not part of your patch so this could be "fixed" in some later patch. It just looks like the conf variable could be clarified it a bit. It looks like a boolean value so we could change the u8_t to bool. This would mean that the use of it would change from 0 -> false and 1 -> true.

finikorg replied:
Original idea about conf is to keep USB selected configuration value (or offset of selected configuration in the conf array)

jukkar commented on subsys/usb/class/netusb/netusb.c:
```diff
@@ -23,13 +23,11 @@
 #include "netusb.h"
```
Could you explain also in commit message why we do this change in this "usb:netusb: Remove multi-function array" patch?

jukkar commented on subsys/usb/class/netusb/function_ecm.c:
```diff
@@ -27,16 +27,12 @@
 struct net_pkt *in_pkt;
 
+/* In a case of low memory skip data to the end of the packet */
+bool skip;
```
and this one too.

jukkar commented on subsys/usb/class/netusb/function_ecm.c:
```diff
@@ -27,16 +27,12 @@
-} ecm;
-
 /* Pointer to pkt assembling at the moment */
 struct net_pkt *in_pkt;
```
Shouldn't this be static?

finikorg replied:
Yes, this is my fault, should be static.

Commit:
usb: netusb: ecm: Make in_pkt static

This reassembly buffer is local.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
usb: netusb: Change 'conf' to 'configured'

conf member represents the configuration state of the device.
Change its type from u8 to bool and clean related functions.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

loicpoulain commented:
- Added comment about multi-function array removing
- Changed conf name and type
- Fix missing static in function_ecm.c

jukkar reviewed (approved):
LGTM

jukkar commented:
@finikorg could you check this too and ack if ok
